man/netxml-ups.8
.TH netxml\-ups 8 "Mon Aug 10 2009" "" "Network UPS Tools (NUT)" 
.SH NAME  
netxml\(hyups \- Driver for Eaton /MGE Network Management Card / Proxy
(XML/HTTP Protocol) equipment
.SH NOTE
This man page only documents the hardware\(hyspecific features of the
netxml\(hyups driver.  For information about the core driver, see  
\fBnutupsdrv\fR(8).

.SH SUPPORTED HARDWARE
netxml\(hyups support all recent Eaton / MGE models which use a Network
Management Card or Proxy (MGE XML/HTTP protocol based). This applies to both
Eaton (previously MGE Office Protection Systems) and to MGE UPS SYSTEMS.
Supported card and proxy models are:

    NMC Minislot (Ref 66102, firmware EA or newer),
    SNMP/Web Minislot card (Ref 66244)
    NMC Transverse (Ref 66074),
    NMC & Modbus/JBus (Ref 66103),
    Network Management Proxy,
    ePDU Monitored (newer version). 

Older models, such as SNMP card (Ref 66062 and Ref 66045), use the SNMP
protocol and should use the snmp\(hyups driver with the "mibs=mge" parameter.

.SH EXTRA ARGUMENTS
This driver supports the following optional settings in the 
\fBups.conf\fR(5):

.IP "timeout=\fIvalue\fR"
The timeout for connecting to and reading from the UPS. Defaults to 5 seconds.
Don't change this value unless you know exactly what you're doing.

This value \fBmust never\fR be higher than half the MAXAGE value specified in
\fBupsd.conf\fR(5), otherwise you risk that \fBupsd\fR(8) declares the driver
stale while it is waiting for a connection to timeout.

.IP "login=\fIvalue\fR"
Set the login value for authenticated mode. This feature also needs the
password argument, and allows value settings in the card.
This feature is not used yet.

.IP "password=\fIvalue\fR"
Set the password value, needed with the login for authenticated mode.
This feature is not used yet.

.SH IMPLEMENTATION
The hostname of the UPS is specified with the "port" value in
\fBups.conf\fR, i.e.:

.nf
	[mgexml]
		driver = netxml\-ups
		port = http://netxml\-ups.example.com:80
.fi

Specifying the method how to connect to the UPS (http, https) is mandatory.
If the port is equal to the default for the method specified (80 for http,
443 for https) it may be omitted.

In order not to overload older NMCs by polling them too frequently, it is
recommended to increase the "pollinterval" (see \fBnutupsdrv\fR(8)) and
\fBups.conf\fR(5)) to at least 5 seconds.

.SH KNOWN ISSUES
Don't connect to the UPS through a proxy. Although it would be trivial to add
support for proxies, this is not recommended and don't ask for it. Not only
because it will prevent the driver to make a persistent connection to the UPS,
but also it adds an additional failure mode. If the proxy goes down (for
whatever reason), the driver will no longer be able to reach the UPS.

.SH AUTHORS
Arjen de Korte <adkorte-guest@alioth.debian.org>

.SH SEE ALSO

.SS The core driver:
\fBnutupsdrv\fR(8)

.SS Internet resources:
The NUT (Network UPS Tools) home page: http://www.networkupstools.org/


NEWS
If you're upgrading from an earlier version, see the UPGRADING file.

---------------------------------------------------------------------------
Release notes for NUT 2.4.1 - what's new since 2.4.0:

 - the microdowell driver has appeared to support various MicroDowell Enterprise
   units (see the "new devices" list below).

 - support for new devices: MicroDowell Enterprise B8, B10, N8, N11, N15, N20,
   N22, N30, N40, N50, N60 and HiBox ST.

 - NUT-Monitor now better handles the ups.status field, and has switched to
   version 1.1.

 - the situation of the build toolchain has been fixed, with regard to the
   "make clean" target and the wrongly removed generated USB files. This brokes
   further configure call. 

---------------------------------------------------------------------------
Release notes for NUT 2.4.0 - what's new since 2.2.2:

 - preliminary support for Power Distribution Units (PDUs): NUT now support
   PDUs, either natively (ie using NUT snmp-ups driver), or through a binding to
   the Powerman daemon. The list of supported PDUs is already quite long,
   including: Eaton ePDUs (Managed and Monitored), some Aphel models, some
   Raritan PDUs, and the whole list of Powerman supported devices:
   http://powerman.sourceforge.net/supported.html
 
 - support for new devices: the various PDUs cited above, Chloride Desk Power
   650, Cyber Power Systems Value 400E/600E/800E (USB models), Delta GES602N,
   Digitus DN-170020, the whole Eaton ranges (mostly composed of MGE Office
   Protection Systems and Powerware units) including BladeUPS, Forza Power
   Technologies SL-1001, HP PowerTrust 2997A, HP R/T 2200 G2, Infosec XP 1000
   and XP 500, Ippon Back Power Pro (serial and USB), Kebo 1200D/D Series,
   Liebert PowerSure Personal XT, MGE Office Protection Systems Protection
   Station, Neus 400va and 600va, Phasak 400VA and 600VA, Plexus 500VA, Powercom
   Black Knight PRO / King PRO and Imperial, PowerKinetics BlackOut Buster,
   Sweex 1000 USB, UNITEK Alpha 500, WinPower CPM-800.

 - NUT now embeds Python client support through the PyNUTClient module and the
   NUT-Monitor application. Both are from David Goncalves, and are still
   available from http://www.lestat.st.
   For more information, refer to scripts/python/README.

 - the dummy-ups driver now support a "repeater" mode. This allows it to act as
   a NUT client, and to forward data. This can be useful for supervision and
   load sharing purposes.

 - tcp-wrappers support has been added to the upsd server, to grant users access
   by source IP for commands that require to be logged into the server. This
   replaces the previous internal implementation (ACL in upsd.conf).

 - the nut.conf file has been introduced to standardize startup configuration
   across the various systems.

 - NUT now ships a bash completion function for 'upsc' command
   (scripts/misc/nut.bash_completion). Simply copy it to /etc/bash_completion.d
 
 - many internal changes to improve maintenability, while lowering the
   maintenance cost (thus allowing developers to focus on what matters: the
   code!). Examples of this are:
   - the USB information automatic extraction to generate the various USB helper
     files,
   - the upsdrv_info_t structure to track more driver information, and remove
     the need for the upsdrv_banner() function
   - common USB code refactoring, as it is done for the serial functions.

 - tons of bugfixes, cleanup and improvements to make NUT stronger than ever!

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.2 - what's new since 2.2.1:

 - support for new devices: APC BACK-UPS XS LCD, Atlantis Land,
   Mustek Powermust Office 650, Oneac XAU models, Powerware PW5115 and
   PW9120 (USB), Nitram Elite 2005
 
 - Integrated Power Management (NUT HAL integration) has reached a
   major milestone: it is now the most advanced UPS integration into
   Power Management layer known in existing OSs. It has received many
   corrections and improvements, and allows to PowerOff the UPS at the
   end of a power cycle (which is the most important feature, not
   supported on other systems).
   The various files are now installed into the correct location.

 - the usbhid-ups driver has received attention. Most notably, the
   shutdown handling has been reworked, and support for MGE UPS SYSTEMS
   3 phases units has been added.

 - snmp-ups now supports MGE* Environment Sensor (ref 66 846).
   The ambient.temperature reporting has also been fixed for units
   other than APC.

 - the netxml-ups driver has appeared to support MGE* network HTTP/XML
   cards.
 
 - NUT now distributes by default the shared version of libupsclient
   (version 1.0.0), and use this for the provided clients (upsmon, upsc,
   upsrw, upscmd). This is part of an effort to reduce NUT's footprint,
   both on disk and in memory.
 
 - powerpanel has reach a new step toward the replacement of nitram and
   cpsups drivers. The final step is scheduled for NUT 2.4.
 
 - many changes, cleanup and fixes to the NUT core and various drivers.

 For a complete list of changes, please refer to the ChangeLog file.

---------------------------------------------------------------------------
Release notes for NUT 2.2.1 - what's new since 2.2.0:

 - support for new devices: all MGE Office Protection Systems units,
   Advice TopGuard 2000, Belkin F6H375-USB, Dynamix UPS1700D, Effekta RM2000MH,
   Jageson Technology Jasuny USPS, Powercom SMK-1500A and SXL-1500A,
   PowerWalker Line-Interactive VI 400/800 and 600, Powerware 9110,
   UNITEK Alpha 2600, UPSonic CXR1000, some vintage serial APC UPSs.

 - the usbhid-ups driver has been improved, and fixed in many areas, through
   a backport of the development (trunk) version.

 - the udev rules, for Linux hotplug support of the USB UPSs, has been
   updated to support kernel newer than 2.6.22.


drivers/netxml-ups.c
/* netxml-ups.c	Driver routines for network XML UPS units 

   Copyright (C)
	2008		Arjen de Korte <adkorte-guest@alioth.debian.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

#include "main.h"
#include "netxml-ups.h"
#include "mge-xml.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <ne_request.h>
#include <ne_basic.h>
#include <ne_props.h>
#include <ne_uri.h>
#include <ne_xmlreq.h>
#include <ne_ssl.h>
#include <ne_auth.h>

#define DRIVER_NAME	"network XML UPS driver"
#define DRIVER_VERSION	"0.21"

/* driver description structure */
upsdrv_info_t	upsdrv_info = {
	DRIVER_NAME,
	DRIVER_VERSION,
	"Arjen de Korte <adkorte-guest@alioth.debian.org>",
	DRV_EXPERIMENTAL,
	{ NULL }
};
/* FIXME:
 * "built with neon library %s" LIBNEON_VERSION 
 * subdrivers (limited to MGE only ATM) */

#define MAXRETRIES	5

/* Global vars */
uint32_t		ups_status = 0;
static int		timeout = 5;
static subdriver_t	*subdriver = &mge_xml_subdriver;
static ne_session	*session = NULL;
static ne_uri		uri;

/* Support functions */
static void netxml_alarm_set(void);
static void netxml_status_set(void);
static int netxml_authenticate(void *userdata, const char *realm, int attempt, char *username, char *password);
static int netxml_dispatch_request(ne_request *request, ne_xml_parser *parser);
static int netxml_get_page(const char *page);

#ifndef HAVE_LIBNEON_SET_CONNECT_TIMEOUT
static void netxml_alarm_handler(int sig)
{
	/* don't do anything here, just return */
}
#endif

void upsdrv_initinfo(void)
{
	char	*page, *last = NULL;
	char	buf[SMALLBUF];
	
	snprintf(buf, sizeof(buf), "%s", subdriver->initinfo);

	for (page = strtok_r(buf, " ", &last); page != NULL; page = strtok_r(NULL, " ", &last)) {

		if (netxml_get_page(page) != NE_OK) {
			continue;
		}

		if (!subdriver->summary) {
			fatalx(EXIT_FAILURE, "%s: XML summary page not found", __func__);
		}

		if (!subdriver->getobject) {
			fatalx(EXIT_FAILURE, "%s: XML getobject page not found", __func__);
		}

		if (!subdriver->setobject) {
			upslogx(LOG_NOTICE, "%s: XML setobject page not found", __func__);
		}

		dstate_setinfo("driver.version.internal", "%s", subdriver->version);

		/* upsh.instcmd = instcmd; */
		/* upsh.setvar = setvar; */

		return;
	}

	fatalx(EXIT_FAILURE, "%s: communication failure [%s]", __func__, ne_get_error(session));
}

void upsdrv_updateinfo(void)
{
	static int	retries = 0, count = 0;
	int		ret;

	if (count++ % 10) {
		ret = netxml_get_page(subdriver->summary);
	} else {
		ret = netxml_get_page(subdriver->getobject);
	}

	if (ret != NE_OK) {
		upsdebugx(1, "%s (%d from %d)", ne_get_error(session), retries, MAXRETRIES);

		if (retries < MAXRETRIES) {
			retries++;
		} else {
			dstate_datastale();
		}

		return;
	}

	retries = 0;

	status_init();

	alarm_init();
	netxml_alarm_set();
	alarm_commit();

	netxml_status_set();
	status_commit();

	dstate_dataok();
}

void upsdrv_shutdown(void)
{
	/* tell the UPS to shut down, then return - DO NOT SLEEP HERE */

	/* maybe try to detect the UPS here, but try a shutdown even if
	   it doesn't respond at first if possible */

	/* replace with a proper shutdown function */
	fatalx(EXIT_FAILURE, "shutdown not supported");

	/* you may have to check the line status since the commands
	   for toggling power are frequently different for OL vs. OB */

	/* OL: this must power cycle the load if possible */

	/* OB: the load must remain off until the power returns */
}

/*
static int instcmd(const char *cmdname, const char *extra)
{
	if (!strcasecmp(cmdname, "test.battery.stop")) {
		ser_send_buf(upsfd, ...);
		return STAT_INSTCMD_HANDLED;
	}

	upslogx(LOG_NOTICE, "%s: unknown command [%s]", __func__, cmdname);
	return STAT_INSTCMD_UNKNOWN;
}
*/

/*
static int setvar(const char *varname, const char *val)
{
	if (!strcasecmp(varname, "ups.test.interval")) {
		ser_send_buf(upsfd, ...);
		return STAT_SET_HANDLED;
	}

	upslogx(LOG_NOTICE, "%s: unknown variable [%s]", __func__, varname);
	return STAT_SET_UNKNOWN;
}
*/

void upsdrv_help(void)
{
}

/* list flags and values that you want to receive via -x */
void upsdrv_makevartable(void)
{
	char	buf[SMALLBUF];

	snprintf(buf, sizeof(buf), "network timeout (default %d seconds)", timeout);
	addvar(VAR_VALUE, "timeout", buf);

	addvar(VAR_VALUE | VAR_SENSITIVE, "login", "login value for authenticated mode");
	addvar(VAR_VALUE | VAR_SENSITIVE, "password", "password value for authenticated mode");
}

void upsdrv_initups(void)
{
	int	ret;
	char	*val;
	FILE	*fp;

#ifndef HAVE_LIBNEON_SET_CONNECT_TIMEOUT
	struct sigaction	sa;

	sigemptyset(&sa.sa_mask);
	sa.sa_flags = 0;

	sa.sa_handler = netxml_alarm_handler;
	sigaction(SIGALRM, &sa, NULL);
#endif
	/* allow override of default network timeout value */
	val = getval("timeout");

	if (val) {
		timeout = atoi(val);

		if (timeout < 1) {
			fatalx(EXIT_FAILURE, "timeout must be greater than 0");
		}
	}

	if (ne_sock_init()) {
		fatalx(EXIT_FAILURE, "%s: failed to initialize socket libraries", progname);
	}

	if (ne_uri_parse(device_path, &uri) || uri.host == NULL) {
		fatalx(EXIT_FAILURE, "%s: invalid hostname '%s'", progname, device_path);
	}
/*
	if (uri.scheme == NULL) {
		uri.scheme = strdup("http");
	}
 
	if (uri.host == NULL) {
		uri.host = strdup(device_path);
	}
 */
	if (uri.port == 0) {
		uri.port = ne_uri_defaultport(uri.scheme);
	}

	upsdebugx(1, "using %s://%s port %d", uri.scheme, uri.host, uri.port);

	session = ne_session_create(uri.scheme, uri.host, uri.port);

	/* timeout if we can't (re)connect to the UPS */
#ifdef HAVE_LIBNEON_SET_CONNECT_TIMEOUT
	ne_set_connect_timeout(session, timeout);
#endif

	/* just wait for a couple of seconds */
	ne_set_read_timeout(session, timeout);

	ne_set_useragent(session, subdriver->version);

	if (strcasecmp(uri.scheme, "https") == 0) {
		ne_ssl_trust_default_ca(session);
	}

	ne_set_server_auth(session, netxml_authenticate, NULL);

	/* if debug level is set, direct output to stderr */
	if (!nut_debug_level) {
		fp = fopen("/dev/null", "w");
	} else {
		fp = stderr;
	}

	if (!fp) {
		fatal_with_errno(EXIT_FAILURE, "Connectivity test failed");
	}

	/* see if we have a connection */
	ret = ne_get(session, subdriver->initups, fileno(fp));

	if (!nut_debug_level) {
		fclose(fp);
	} else {
		fprintf(fp, "\n");
	}

	if (ret != NE_OK) {
		fatalx(EXIT_FAILURE, "Connectivity test: %s", ne_get_error(session));
	}

	upslogx(LOG_INFO, "Connectivity test: %s", ne_get_error(session));
}

void upsdrv_cleanup(void)
{
	free(subdriver->getobject);
	free(subdriver->setobject);

	if (session) {
		ne_session_destroy(session);
	}

	ne_uri_free(&uri);
}

/**********************************************************************
 * Support functions
 *********************************************************************/

static int netxml_get_page(const char *page)
{
	int		ret;
	ne_request	*request;
	ne_xml_parser	*parser;

	upsdebugx(3, "%s: %s", __func__, page);

	if (strlen(page) < 1) {
		ne_set_error(session, "Attempt to read from NULL page");
		return NE_ERROR;
	}

	request = ne_request_create(session, "GET", page);

	parser = ne_xml_create();

	ne_xml_push_handler(parser, subdriver->startelm_cb, subdriver->cdata_cb, subdriver->endelm_cb, NULL);

	ret = netxml_dispatch_request(request, parser);

	ne_xml_destroy(parser);
	ne_request_destroy(request);

	upsdebugx(3, "%s: %s", __func__, ne_get_error(session));

	return ret;
}

static int netxml_dispatch_request(ne_request *request, ne_xml_parser *parser)
{
	int ret;

#ifdef HAVE_LIBNEON_SET_CONNECT_TIMEOUT
	/*
	 * Starting with neon-0.27.0 the ne_xml_dispatch_request() function will check
	 * for a valid XML content-type (following RFC 3023 rules) in the header.
	 * Unfortunately, (at least) the Transverse NMC doesn't follow this RFC, so
	 * we can't use this anymore and we'll have to roll our own here.
	 */
	do {
		ret = ne_begin_request(request);

		if (ret != NE_OK) {
			break;
		}

		ret = ne_xml_parse_response(request, parser);

		if (ret == NE_OK) {
			ret = ne_end_request(request);
		}

	} while (ret == NE_RETRY);

#else
	/*
	 * Up to neon-0.27.0 this library would connect() in blocking mode. We
	 * don't want that, so we interrupt it with an alarm if it takes longer
	 * than we think is reasonable so that we don't have to wait until it
	 * finally times out (which may take several minutes).
	 */
	alarm(timeout+1);

	ret = ne_xml_dispatch_request(request, parser);

 	alarm(0);
#endif

	return ret;
}

/* Supply the 'login' and 'password' when authentication is required */
static int netxml_authenticate(void *userdata, const char *realm, int attempt, char *username, char *password)
{
	char	*val;

	upsdebugx(2, "%s: realm = [%s], attempt = %d", __func__, realm, attempt);

	val = getval("login");
	snprintf(username, NE_ABUFSIZ, "%s", val ? val : "");

	val = getval("password");
	snprintf(password, NE_ABUFSIZ, "%s", val ? val : "");

	return attempt;
}

/* Convert the local status information to NUT format and set NUT
   alarms. */
static void netxml_alarm_set(void)
{
	if (STATUS_BIT(REPLACEBATT)) {
		alarm_set("Replace battery!");
	}
	if (STATUS_BIT(SHUTDOWNIMM)) {
		alarm_set("Shutdown imminent!");
	}
	if (STATUS_BIT(FANFAIL)) {
		alarm_set("Fan failure!");
	}
	if (STATUS_BIT(NOBATTERY)) {
		alarm_set("No battery installed!");
	}
	if (STATUS_BIT(BATTVOLTLO)) {
		alarm_set("Battery voltage too low!");
	}
	if (STATUS_BIT(BATTVOLTHI)) {
		alarm_set("Battery voltage too high!");
	}
	if (STATUS_BIT(CHARGERFAIL)) {
		alarm_set("Battery charger fail!");
	}
	if (STATUS_BIT(OVERHEAT)) {
		alarm_set("Temperature too high!");
	}
	if (STATUS_BIT(COMMFAULT)) {
		alarm_set("Communication fault!");
	}
	if (STATUS_BIT(INTERNALFAULT)) {
		alarm_set("Internal UPS fault!");
	}
	if (STATUS_BIT(FUSEFAULT)) {
		alarm_set("Fuse fault!");
	}
	if (STATUS_BIT(BYPASSAUTO)) {
		alarm_set("Automatic bypass mode!");
	}
	if (STATUS_BIT(BYPASSMAN)) {
		alarm_set("Manual bypass mode!");
	}
}

/* Convert the local status information to NUT format and set NUT
   status. */
static void netxml_status_set(void)
{
	if (STATUS_BIT(VRANGE)) {
		dstate_setinfo("input.transfer.reason", "input voltage out of range");
	} else if (STATUS_BIT(FRANGE)) {
		dstate_setinfo("input.transfer.reason", "input frequency out of range");
	} else {
		dstate_delinfo("input.transfer.reason");
	}

	if (STATUS_BIT(ONLINE)) {
		status_set("OL");		/* on line */
	} else {
		status_set("OB");		/* on battery */
	}
	if (STATUS_BIT(DISCHRG)) {
		status_set("DISCHRG");		/* discharging */
	}
	if (STATUS_BIT(CHRG)) {
		status_set("CHRG");		/* charging */
	}
	if (STATUS_BIT(LOWBATT) || STATUS_BIT(SHUTDOWNIMM)) {
		status_set("LB");		/* low battery */
	}
	if (STATUS_BIT(OVERLOAD)) {
		status_set("OVER");		/* overload */
	}
	if (STATUS_BIT(REPLACEBATT)) {
		status_set("RB");		/* replace batt */
	}
	if (STATUS_BIT(TRIM)) {
		status_set("TRIM");		/* SmartTrim */
	}
	if (STATUS_BIT(BOOST)) {
		status_set("BOOST");		/* SmartBoost */
	}
	if (STATUS_BIT(BYPASSAUTO) || STATUS_BIT(BYPASSMAN)) {
		status_set("BYPASS");		/* on bypass */
	}
	if (STATUS_BIT(OFF)) {
		status_set("OFF");		/* ups is off */
	}
}
